external/llvm-project/mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp

```diff
@@ -830,6 +830,178 @@ LogicalResult PackedStochRoundFp8OpLowering::matchAndRewrite(
+        "quad_perm",  "row_shl",         "row_shr",      "row_ror",
+        "row_mirror", "row_half_mirror", "wave_shl",     "wave_shr",
+        "wave_ror",   "wave_rol",        "row_bcast_15", "row_bcast_31"};
+    int attrCount = 0;
+    for (const auto &attrName : attributeNames) {
+      if (auto attr = op->getAttr(attrName)) {
+        int attrValue = 0;
+        if (attrName == "row_shl" || attrName == "row_shr" ||
+            attrName == "row_ror") {
+          auto intAttr = attr.dyn_cast<IntegerAttr>();
+          // If attribute is IntegerAttr, use its value for dppCtrl
+          attrValue = intAttr.getInt();
+          if (attrValue < 1 || attrValue > 15) {
+            return op->emitError(
+                       "Attribute value must be between 1 and 15, but got ")
+                   << attrValue;
+          }
+          if (attrName == "row_shl")
+            attrValue += DppCtrl::ROW_SHL0;
+          else if (attrName == "row_shr")
+            attrValue += DppCtrl::ROW_SHR0;
+          else if (attrName == "row_ror")
+            attrValue += DppCtrl::ROW_ROR0;
+        }
+
+        else if (attrName == "row_bcast_15" || attrName == "row_bcast_31") {
+          if (attrName == "row_bcast_15")
+            attrValue = DppCtrl::BCAST15;
+          else if (attrName == "row_bcast_31")
+            attrValue = DppCtrl::BCAST31;
+        }
+
+        else if (attrName == "wave_shl" || attrName == "wave_shr" ||
+                 attrName == "wave_ror" || attrName == "wave_rol" ||
+                 attrName == "row_mirror" || attrName == "row_half_mirror") {
+          auto boolAttr = attr.dyn_cast<BoolAttr>();
+          if (!boolAttr.getValue()) {
+            return op->emitError("Attribute must be true ") << "\n";
+          }
+
+          if (attrName == "wave_shl")
+            attrValue = DppCtrl::WAVE_SHL1;
+          else if (attrName == "wave_shr")
+            attrValue = DppCtrl::WAVE_SHR1;
+          else if (attrName == "wave_ror")
+            attrValue = DppCtrl::WAVE_ROR1;
+          else if (attrName == "wave_rol")
+            attrValue = DppCtrl::WAVE_ROL1;
+          else if (attrName == "row_mirror")
+            attrValue = DppCtrl::ROW_MIRROR;
+          else if (attrName == "row_half_mirror")
+            attrValue = DppCtrl::ROW_HALF_MIRROR;
+        }
+
+        else if (attrName == "quad_perm") {
+          auto quadPermAttr = op->getAttrOfType<ArrayAttr>("quad_perm");
+          for (int i = 3; i >= 0; --i) {
+            auto intAttr = quadPermAttr[i].dyn_cast<IntegerAttr>();
+            int num = intAttr.getInt();
+            if (num > 3) {
+              return op->emitError(
+                  "Invalid value for quad_perm, must be in the range 0 to 3.");
+            }
+            attrValue |= num << (i * 2);
+          }
+        }
+        dppCtrl =
+            rewriter.create<LLVM::ConstantOp>(loc, llvmI32Type, attrValue);
+        attrCount++;
+      }
+    }
+
+    if (attrCount != 1) {
```

Comment:
This check is why I wanted to do a DPP kind enum. Also, this sort of testing (or, with the change I want, checking of whether `permArgument` is the right type) should be in the operation verifier, not all the way down in the lowering.